Source/SwiftLintCore/Models/Baseline.swift

```diff
@@ -0,0 +1,184 @@
+            for (key, ruleViolations) in groupedRuleViolations {
+                guard let baselineViolations = groupedBaselineViolations[key] else {
+                    filteredViolations.formUnion(ruleViolations)
+                    continue
+                }
+                if ruleViolations.count > baselineViolations.count {
+                    filteredViolations.formUnion(ruleViolations)
+                }
+            }
+        }
+
+        let violationsWithAbsolutePaths = Set(filteredViolations.violationsWithAbsolutePaths)
+        return violations.filter { violationsWithAbsolutePaths.contains($0) }
+    }
+
+    /// Returns the violations that are present in another `Baseline`, but not in this one.
+    ///
+    /// The violations are filtered using the same algorithm as the `filter` method above.
+    ///
+    /// - parameter otherBaseline: The other `Baseline`.
+    public func compare(_ otherBaseline: Baseline) -> [StyleViolation] {
+        otherBaseline.baselineViolations.flatMap {
+            filter($1.violationsWithAbsolutePaths)
+        }
+    }
+}
+
+private extension Sequence where Element == StyleViolation {
+    var baselineViolations: [BaselineViolation] {
+        var lineCache = LineCache()
+        return map {
+            $0.baselineViolation(text: lineCache.text(at: $0.location))
+        }
+    }
+}
+
+private struct LineCache {
+    private var lines: [String: [String]] = [:]
+
+    mutating func text(at location: Location) -> String {
+        let lineNumber = (location.line ?? 0) - 1
+        guard lineNumber > 0, let absolutePath = location.file else {
+            return ""
+        }
+        if let fileLines = lines[absolutePath] {
+            return (fileLines.isNotEmpty && lineNumber < fileLines.count) ? fileLines[lineNumber] : ""
+        }
+        if let fileLines = SwiftLintFile(path: absolutePath)?.lines.map(\.content) {
+            lines[absolutePath] = fileLines
+            return lineNumber < fileLines.count ? fileLines[lineNumber] : ""
+        }
+        return ""
+    }
+}
+
+private extension Sequence where Element == BaselineViolation {
+    var violationsWithAbsolutePaths: [StyleViolation] {
+        map { $0.violation.withAbsolutePath }
+    }
+
+    func groupedByFile() -> ViolationsPerFile {
+        Dictionary(grouping: self) { $0.violation.location.relativeFile ?? "" }
+    }
+
+    func groupedByRuleIdentifier() -> ViolationsPerRule {
+        Dictionary(grouping: self, by: \.violation.ruleIdentifier)
+    }
+
+    func groupedByRuleIdentifier(filteredBy existingViolations: [BaselineViolation]) -> ViolationsPerRule {
+        Set(self).subtracting(existingViolations).groupedByRuleIdentifier()
+    }
+}
+
+private extension StyleViolation {
+    var withAbsolutePath: StyleViolation {
+        let absolutePath: String? = if let relativePath = location.file {
+            FileManager.default.currentDirectoryPath + "/" + relativePath
+        } else {
+            nil
+        }
```

Comment:
Always unclear how to format this. Better?
```suggestion
        let absolutePath: String? =
            if let relativePath = location.file {
                FileManager.default.currentDirectoryPath + "/" + relativePath
            } else {
                nil
            }
```

Reply:
yep, I think that makes it clearer